File: hooks/continuum_session_capture.py
#!/usr/bin/env python3
"""SessionEnd hook: hand the finished transcript to Continuum, which extracts
memory candidates for review in the dashboard. Nothing is written to memory
automatically — the candidates sit in a queue until approved.

Installed by Continuum's install-hook script. Pure stdlib and no shell, so the
same file runs unchanged on Linux, macOS and Windows. Exits 0 on any failure.
"""

import json
import os
import sys
import urllib.request
from pathlib import Path

STATE_DIR = Path.home() / ".continuum"
TOKEN_FILE = STATE_DIR / "hook-token"
DISABLE_FILE = STATE_DIR / "hook-disabled"
CAPTURE_DISABLE_FILE = STATE_DIR / "capture-disabled"
WORKSPACE_MAP = STATE_DIR / "workspace-map.json"

CONTINUUM_URL = os.environ.get("CONTINUUM_MCP_URL", "https://continuum-mcp.sshogunn.org")
TIMEOUT_SECONDS = 10.0
TURN_CHARS = 4000
TRANSCRIPT_CHARS = 40000


def text_of(message):
    content = message.get("content")
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        return " ".join(
            block.get("text", "")
            for block in content
            if isinstance(block, dict) and block.get("type") == "text"
        )
    return ""


def transcript_text(transcript_path):
    turns = []
    with open(transcript_path, encoding="utf-8", errors="replace") as f:
        for line in f:
            try:
                entry = json.loads(line)
            except ValueError:
                continue
            message = entry.get("message")
            if not isinstance(message, dict):
                continue
            role = message.get("role")
            if role not in ("user", "assistant"):
                continue
            body = text_of(message).strip()
            if body:
                turns.append("%s: %s" % (role, body[:TURN_CHARS]))
    return "\n\n".join(turns)[-TRANSCRIPT_CHARS:]


def workspace_for(cwd):
    if not cwd:
        return ""
    try:
        with open(WORKSPACE_MAP, encoding="utf-8") as f:
            return json.load(f).get(cwd, "") or ""
    except (OSError, ValueError):
        return ""


def main():
    if DISABLE_FILE.exists() or CAPTURE_DISABLE_FILE.exists():
        return
    try:
        token = TOKEN_FILE.read_text(encoding="utf-8").strip()
    except OSError:
        return
    if not token:
        return

    data = json.load(sys.stdin)
    transcript_path = data.get("transcript_path") or ""
    if not transcript_path:
        return
    transcript = transcript_text(transcript_path)
    if not transcript.strip():
        return

    payload = {"transcript": transcript, "session_id": data.get("session_id") or ""}
    workspace = workspace_for(data.get("cwd") or "")
    if workspace:
        payload["workspace"] = workspace

    request = urllib.request.Request(
        CONTINUUM_URL.rstrip("/") + "/hook/session",
        data=json.dumps(payload).encode("utf-8"),
        headers={
            "Authorization": "Bearer " + token,
            "Content-Type": "application/json",
        },
        method="POST",
    )
    with urllib.request.urlopen(request, timeout=TIMEOUT_SECONDS) as response:
        response.read()


if __name__ == "__main__":
    try:
        sys.stdin.reconfigure(encoding="utf-8", errors="replace")
    except (AttributeError, ValueError):
        pass
    try:
        main()
    except Exception:
        pass
    sys.exit(0)

File: hooks/continuum_session_capture.js
#!/usr/bin/env node
"use strict";

// SessionEnd hook: hand the finished transcript to Continuum, which extracts
// memory candidates for review in the dashboard. Nothing is written to memory
// automatically — the candidates sit in a queue until approved.
//
// Installed by Continuum's install-hook script. Pure Node core modules and no
// dependencies, so the same file runs unchanged on Linux, macOS and Windows.
// Exits 0 on any failure.

const fs = require("fs");
const os = require("os");
const path = require("path");

const STATE_DIR = path.join(os.homedir(), ".continuum");
const TOKEN_FILE = path.join(STATE_DIR, "hook-token");
const DISABLE_FILE = path.join(STATE_DIR, "hook-disabled");
const CAPTURE_DISABLE_FILE = path.join(STATE_DIR, "capture-disabled");
const WORKSPACE_MAP = path.join(STATE_DIR, "workspace-map.json");

const CONTINUUM_URL = process.env.CONTINUUM_MCP_URL || "https://continuum-mcp.sshogunn.org";
const TIMEOUT_MS = 10000;
const TURN_CHARS = 4000;
const TRANSCRIPT_CHARS = 40000;

function textOf(message) {
  const content = message.content;
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .filter((block) => block && typeof block === "object" && block.type === "text")
      .map((block) => block.text || "")
      .join(" ");
  }
  return "";
}

function transcriptText(transcriptPath) {
  const raw = fs.readFileSync(transcriptPath, "utf-8");
  const turns = [];
  for (const line of raw.split("\n")) {
    if (!line.trim()) continue;
    let entry;
    try {
      entry = JSON.parse(line);
    } catch {
      continue;
    }
    const message = entry.message;
    if (!message || typeof message !== "object") continue;
    const role = message.role;
    if (role !== "user" && role !== "assistant") continue;
    const body = textOf(message).trim();
    if (body) turns.push(`${role}: ${body.slice(0, TURN_CHARS)}`);
  }
  return turns.join("\n\n").slice(-TRANSCRIPT_CHARS);
}

function workspaceFor(cwd) {
  if (!cwd) return "";
  try {
    return JSON.parse(fs.readFileSync(WORKSPACE_MAP, "utf-8"))[cwd] || "";
  } catch {
    return "";
  }
}

function readStdin() {
  return new Promise((resolve, reject) => {
    let data = "";
    process.stdin.setEncoding("utf-8");
    process.stdin.on("data", (chunk) => (data += chunk));
    process.stdin.on("end", () => resolve(data));
    process.stdin.on("error", reject);
  });
}

async function main() {
  if (fs.existsSync(DISABLE_FILE) || fs.existsSync(CAPTURE_DISABLE_FILE)) return;
  let token;
  try {
    token = fs.readFileSync(TOKEN_FILE, "utf-8").trim();
  } catch {
    return;
  }
  if (!token) return;

  let data;
  try {
    data = JSON.parse(await readStdin());
  } catch {
    return;
  }
  const transcriptPath = data.transcript_path || "";
  if (!transcriptPath) return;
  const transcript = transcriptText(transcriptPath);
  if (!transcript.trim()) return;

  const payload = { transcript, session_id: data.session_id || "" };
  const workspace = workspaceFor(data.cwd || "");
  if (workspace) payload.workspace = workspace;

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
  try {
    await fetch(`${CONTINUUM_URL.replace(/\/$/, "")}/hook/session`, {
      method: "POST",
      headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
      body: JSON.stringify(payload),
      signal: controller.signal,
    });
  } finally {
    clearTimeout(timer);
  }
}

(async () => {
  try {
    await main();
  } catch {
    // exits 0 regardless
  }
  process.exit(0);
})();
